precompiles/gmp/src/lib.rs

```diff
@@ -0,0 +1,317 @@
+			32,
+			&mut transfer_with_payload.payload.as_bytes(),
+		)
+		.map_err(|_| RevertReason::Custom("Invalid GMP Payload".into()))?;
+		log::debug!(target: "gmp-precompile", "user action: {:?}", user_action);
+
+		// inspect the token the user wants to use: make sure it is XCM-capable
+		let asset_address: H160 = transfer_with_payload
+			.token_address
+			.try_into()
+			.map_err(|_| revert("Asset address is not a H160"))?;
+		let currency_account_id = Runtime::AddressMapping::into_account_id(wrapped_address.into());
+
+		let currency_id: <Runtime as orml_xtokens::Config>::CurrencyId =
+			Runtime::account_to_currency_id(currency_account_id)
+				.ok_or(revert("Unsupported asset, not a valid currency id"))?;
+
+		let amount = transfer_with_payload
+			.amount
+			.try_into()
+			.map_err(|_| revert("Amount overflows balance"))?;
+
+		log::debug!(target: "gmp-precompile", "attempt to transfer {:?} of {:?}", amount, currency_id);
+
+		// TODO: now check before balance
+
+		// Complete a "Contract Controlled Transfer" with the given Wormhole VAA.
+		// We need to invoke Wormhole's completeTransferWithPayload function, passing it the VAA,
+		// then use the returned payload to decide what to do.
+		let sub_context = Context {
+			caller: handle.code_address(), // TODO: can we trust this to always be "this precompile"?
+			address: wormhole_bridge,
+			apparent_value: U256::zero(), // TODO: any reason to pass value on, or reject txns with value?
+		};
+
+		log::debug!(target: "gmp-precompile", "calling Wormhole completeTransferWithPayload on {}...", wormhole_bridge);
+		let (reason, output) = handle.call(
+			wormhole_bridge,
+			None,
+			EvmDataWriter::new_with_selector(COMPLETE_TRANSFER_WITH_PAYLOAD_SELECTOR)
+				.write(wormhole_vaa)
+				.build(),
+			handle.gas_limit(), // TODO
+			false,
+			&sub_context,
+		);
+
+		ensure_exit_reason_success(reason, &output[..])?;
+
+		// query our "after" balance (our being this precompile)
+		let output = Self::call(
+			handle,
+			wrapped_address.into(),
+			EvmDataWriter::new_with_selector(BALANCE_OF_SELECTOR)
+				.write(Address::from(handle.code_address()))
+				.build(),
+		)?;
+		let mut reader = EvmDataReader::new(&output[..]);
+		let after_amount: U256 = reader.read()?;
+		log::debug!(target: "gmp-precompile", "after balance: {}", after_amount);
+
+		let amount_transfered = after_amount.saturating_sub(before_amount);
+
+		// TODO: review
+		if amount_transfered < transfer_with_payload.amount {
+			log::debug!(target: "gmp-precompile", "insufficient funds bridged ({} < {})", amount_transfered, transfer_with_payload.amount);
+			return Err(PrecompileFailure::Revert {
+				exit_status: ExitRevert::Reverted,
+				output: "insufficient funds bridged".into(),
+			});
+		}
+
+		// TODO: Wormhole might have transfered unsupported tokens; we should handle this case
+		//       gracefully (maybe that's as simple as reverting)
+
+		// TODO:
+		let weight_limit: u64 = 1_000_000_000_000u64;
```

Comment:
For this initial version, I'd rather not force the user to deal with this. Is this value sufficient for most simple xtokens transfers?

In the future, I think we could either require this and let dapp designers figure it out (e.g. provide their own sane defaults) or make it optional and provide our own sane defaults.

Reply:
It depend on the implementation in the destination chain, so we can't provide a sane defaults. The value should be provided by the dapp, and in all dimensions (ref time and proof size).

Reply:
We can define the weight limit as optional, and use the Unlimited variant is nothing was provided

Reply:
That would need to be included in the payload with the multilocation or would this be provided with the VAA when calling the precompile ?

Reply:
In the payload

Reply:
Addressed in 39c61e6 by using `Unlimited`. This should be improved when we work on the fee mechanism, I suppose...